### PR TITLE
Bloc 128

### PR DESCRIPTION
**Pull request by Andrei-Masilevich**
Initial registration pool functionality

### CODE REVIEW

File: libraries/chain/dbs_registration_committee.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+namespace scorum {
+namespace chain {
+
+dbs_registration_committee::dbs_registration_committee(database& db)
+    : _base_type(db)
+{
+}
+
+bool dbs_registration_committee::is_committee_exists() const
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_committee.cpp]: Remove excessive method

File: libraries/chain/dbs_registration_committee.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+{
+    registration_committee_member_refs_type ret;
+
+    const auto &idx = db_impl().get_index<registration_committee_member_index>().indicies();
+    for (auto it = idx.cbegin(); it != idx.cend(); ++it)
+    {
+        ret.push_back(std::cref(*it));
+    }
+
+    FC_ASSERT(!ret.empty(), "Committee is empty.");
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_committee.cpp]: Remove Assert, you have this check in create_committee()

File: libraries/chain/dbs_account.cpp [hunk 3]
```diff
@@ -52,14 +52,42 @@ void dbs_account::check_account_existence(const account_authority_map& names,
...
 
-const account_object& dbs_account::create_account_by_faucets(const account_name_type& new_account_name,
-                                                             const account_name_type& creator_name,
-                                                             const public_key_type& memo_key,
-                                                             const string& json_metadata,
-                                                             const authority& owner,
-                                                             const authority& active,
-                                                             const authority& posting,
-                                                             const asset& fee)
+const account_object& dbs_account::create_initial_account(const account_name_type& new_account_name,
```
[the remark this answers is not in the document]
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_account.cpp]: It is for db_genesis. Usage lost while merge. Add to db_genesis please

File: libraries/chain/dbs_registration_committee.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        FC_ASSERT(false, "Can't initialize at least one member.");
+    }
+
+    return get_committee();
+}
+
+const registration_committee_member_object&
+dbs_registration_committee::add_member(const account_name_type& account_name)
+{
+    FC_ASSERT(!get_committee().empty(), "No committee to add memeber.");
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_committee.cpp]: Why we can't add new member if committee has no members yet?

File: libraries/chain/dbs_registration_committee.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+{
+    db_impl().modify(member, [&](registration_committee_member_object& m) { modifier(m); });
+}
+
+uint64_t dbs_registration_committee::_get_member_count() const
+{
+    return db_impl().get_index<registration_committee_member_index>().indicies().size();
+}
+
+bool dbs_registration_committee::_check_member_exist(const account_name_type& account_name) const
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_committee.cpp]: Just _member_exist() name for function is more comprehensible

File: libraries/chain/dbs_registration_pool.cpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+namespace scorum {
+namespace chain {
+
+dbs_registration_pool::dbs_registration_pool(database& db)
+    : _base_type(db)
+{
+}
+
+bool dbs_registration_pool::is_pool_exists() const
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: Do we really need this public method?

File: libraries/chain/dbs_registration_pool.cpp [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    FC_ASSERT(SCORUM_REGISTRATION_BONUS_LIMIT_PER_MEMBER_PER_N_BLOCK_AMOUNT > 0, "Invalid ${1} value.",
+              ("1", SCORUM_REGISTRATION_BONUS_LIMIT_PER_MEMBER_PER_N_BLOCK_AMOUNT));
+
+    const dbs_account& account_service = db().obtain_service<dbs_account>();
+
+    account_service.check_account_existence(member_name);
+
+    const registration_pool_object& this_pool = get_pool();
+
+    share_type per_reg = _calculate_per_reg(this_pool);
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: Why not to use asset instead of share_type?

File: libraries/chain/dbs_registration_pool.cpp [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+    return asset(per_reg, VESTS_SYMBOL);
+}
+
+share_type dbs_registration_pool::_decrease_balance(const registration_pool_object& this_pool,
+                                                    const share_type& balance)
+{
+    FC_ASSERT(balance > 0, "Invalid balance.");
+
+    share_type ret{ 0 };
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: asset?

File: libraries/chain/dbs_registration_pool.cpp (hunk 7 above)
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: Why not to use asset instead of share_type?

File: libraries/chain/dbs_registration_pool.cpp [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+bool dbs_registration_pool::is_pool_exists() const
+{
+    return !db_impl().get_index<registration_pool_index>().indicies().empty();
+}
+
+const registration_pool_object& dbs_registration_pool::get_pool() const
+{
+    const auto& idx = db_impl().get_index<registration_pool_index>().indicies();
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: May be const auto& idx = db_impl().get<registration_pool>(); is better?

File: libraries/chain/dbs_registration_pool.cpp [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        {
+            ret = pool.balance.amount;
+            pool.balance.amount = 0;
+        }
+    });
+
+    return ret;
+}
+
+share_type dbs_registration_pool::_calculate_per_reg(const registration_pool_object& this_pool)
```
Comment (abolinbsu) [inline libraries/chain/dbs_registration_pool.cpp]: I think this_pool is excessive parameter, you have just one pool object.

File: libraries/chain/include/scorum/chain/genesis_state.hpp [hunk 11]
```diff
@@ -30,21 +30,22 @@ struct genesis_state_type
...
-        : init_supply(supply)
-    {
-    }
+        uint8_t stage;
+        uint32_t users;
+        uint16_t bonus_percent;
+    };
 
-    sp::share_type init_supply;
+    sp::share_type registration_supply = 0;
```
Comment (abolinbsu) [inline libraries/chain/include/scorum/chain/genesis_state.hpp]: assets?

File: libraries/chain/dbs_registration_pool.cpp (hunk 6 above)
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_registration_pool.cpp]: Maybe. Let it be.

File: libraries/chain/dbs_registration_committee.cpp (hunk 4 above)
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_registration_committee.cpp]: But who will vote?

File: libraries/chain/dbs_registration_committee.cpp (hunk 5 above)
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_registration_committee.cpp]: I don't think sow

File: genesis.json [hunk 12]
```diff
@@ -30,5 +30,29 @@
 	}, {
 		"owner_name": "bob",
 		"block_signing_key": "SCR5gSUmAuNo1ry9cv88o3JmM7K6iJJtTh3f2NyMzWrKa6ZBz365B"
+	}],
+
+	"init_rewards_supply": "0.100 TESTS",
```
Comment (alexplevako) [inline genesis.json]: 0.100 SCORUM

File: genesis.json [hunk 13]
```diff
@@ -30,5 +30,29 @@
...
+
+	"registration_committee": ["alice", "bob"],
+	"registration_supply": "1.000 TESTS",
```
Comment (alexplevako) [inline genesis.json]: SCORUM not tests

File: tests/common/database_fixture.cpp [hunk 14]
```diff
@@ -475,6 +462,51 @@ void database_fixture::validate_database(void)
...
+        genesis_state.registration_maximum_bonus.amount /= 2;
+    }
+
+    if (!genesis_state.registration_supply.amount.value)
+    {
+        genesis_state.registration_supply = SCORUM_REGISTRATION_BONUS_LIMIT_PER_MEMBER_PER_N_BLOCK;
+        genesis_state.registration_supply.amount /= 2;
+    }
+
+    private_key_type init_delegate_priv_key = private_key_type::regenerate(fc::sha256::hash(string("init_key")));
```
Comment (alexplevako) [inline tests/common/database_fixture.cpp]: break large functions into small ones. This code was in the function ```create_initdelegate_for_genesis_state```.

**Pull request merged**